Add i.MX8M Mini SoC and board support

Mani-Sadhasivam opened the pull request:
Hello,

This PR adds SoC and board support for NXP i.MX8M Mini SoC. NXP has a wide variety of i.MX8 series SoCs with different configurations. So I'm not exactly sure how to model the SoC hierarchy here. But what I have done is to create a separate `mimx8mm_m4` SoC directory for i.MX8M Mini SoCs.

And the board support is added for the i.MX8M Mini LPDDR EVK board. The model name for this board is `8MMINILPD4-EVK` but since we can't use that name directly, I have added it as `mimx8mm_evk`.

The corresponding PR for adding SDK support to NXP module is here: https://github.com/zephyrproject-rtos/hal_nxp/pull/16

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Fixes #17814

Mani-Sadhasivam committed:
dts: arm: nxp: Add i.MX8M Mini devicetree support

Add devicetree support for NXP i.MX8M Mini SoC for utilizing cortex M4
core.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
soc: arm: nxp_imx: Add i.MX8M Mini SoC support

Add SoC support for the NXP i.MX8M Mini series MIMX8MM6 SoC. This SoC
has a quad Cortex-A53 cluster and a single core Cortex-M4 core. Zephyr
support is added to the Cortex-M4 core for running at 800MHz.

More information about the SoC can be found here: https://www.nxp.com/products/processors-and-microcontrollers/arm-processors/i.mx-applications-processors/i.mx-8-processors/i.mx-8m-mini-arm-cortex-a53-cortex-m4-audio-voice-video:i.MX8MMINI

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
drivers: serial: Add NXP IUART driver

Add IUART driver based on MCUX SDK. This driver is used to provide
serial console support on i.MX8M Mini SoC.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
drivers: clock_control: Add support for MCUX CCM IUART clock

Add support for controlling the MCUX CCM IUART clock.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
boards: arm: Add NXP i.MX8M Mini EVK board support

Add board support for NXP i.MX8M Mini EVK. This board has the following
features:

Processor    : i.MX8M Mini Quad applications processor
Memory       : 32-bit LPDDR4 w/2 GB
               eMMC 5.0/5.1 w/16 GB
               SD/MMC connector
               QSPI w/32 MB
Connectivity : MIMO 1x1 Wi-Fi 802.11a/b/g/n/ac and BT4.1
               Ethernet
               PCIe M.2
USB          : 2x USB 3.0 Type C
Debug        : JTAG connector
               MicroUSB for debug console

More information about this board can be found in NXP website: https://www.nxp.com/design/development-boards/i.mx-evaluation-and-development-boards/evaluation-kit-for-thebr-i.mx-8m-mini-applications-processor:8MMINILPD4-EVK

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
west: Update NXP HAL

Update NXP HAL to use the i.MX8M SDK support.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam commented:
recheck

dbkinder commented on boards/arm/mimx8mm_evk/doc/index.rst:
```diff
@@ -0,0 +1,185 @@
+|           |            | serial port-interrupt               |
++-----------+------------+-------------------------------------+
+
+The default configuration can be found in the defconfig file:
```
Renders better as:
```
The default configuration can be found in the defconfig file:
:zephyr_file:`boards/arm/mimx8mm_evk/mimx8mm_evk_defconfig`.
```

dbkinder commented on boards/arm/mimx8mm_evk/doc/index.rst:
```diff
@@ -0,0 +1,185 @@
+   - &ocram_sys
+   - &ocram_s_sys
+
+Below you will find the instructions to load and run Zephyr on M4 from
```
Better would be:
```
Load and run Zephyr on M4 from A53 using u-boot by
copying the compiled ``zephyr.bin`` to the first FAT partition of the SD card and
plug the SD card into the board. Power it up and stop the u-boot execution.

Load the M4 binary onto the desired memory and
start its execution using:

.. code-block:: console

   fatload mmc 0:1 0x7e0000 zephyr.bin;bootaux 0x7e0000
```

carlescufi commented:
@Mani-Sadhasivam can you rebase and address comments? If you want this in 2.1 needs to be done before tomorrow.

Mani-Sadhasivam commented:
> @Mani-Sadhasivam can you rebase and address comments? If you want this in 2.1 needs to be done before tomorrow.

@carlescufi Will do. Fyi, still there is no review done for the code.

Mani-Sadhasivam commented:
@MaureenHelm I think the CI failure is not related to this PR or the corresponding HAL update. I can reproduce the build failure locally with master branch for those boards.

dbkinder reviewed (approved):
doc changes LGTM, thanks!

MaureenHelm commented on soc/arm/nxp_imx/mimx8mm6_m4/soc.c:
```diff
@@ -0,0 +1,147 @@
+};
+
+/* AUDIO PLL2 configuration */
+const ccm_analog_frac_pll_config_t g_audioPll2Config = {
```
make these configs and all the functions in this file `static`

Mani-Sadhasivam replied:
okay

MaureenHelm commented on boards/arm/mimx8mm_evk/doc/index.rst:
```diff
@@ -0,0 +1,182 @@
+
+.. code-block:: console
+
+   JLinkGDBServer -device MIMX8MD6_M4
```
instead of manually starting the JLinkGDBServer, can you use `west debug` or `west attach`? you'll need to add a `board.cmake` file (see other imx rt or kinetis boards for an example)

Mani-Sadhasivam replied:
Okay.

MaureenHelm reviewed (changes requested):
Thanks @Mani-Sadhasivam!

Mani-Sadhasivam committed:
CODEOWNERS: Add myself as the codeowner for imx8m evk and uart driver

Add myself as the codeowner for imx8m evk and uart driver.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam commented:
@MaureenHelm Comments incorporated!

MaureenHelm reviewed (changes requested):
Please fix the shippable failures

Mani-Sadhasivam commented:
@MaureenHelm Fixed all CI failures!

Mani-Sadhasivam committed:
modules: Add RDC module for MCUX family

RDC module is used by i.MX8MM SoC, so let's create a Kconfig symbol
which will be used to enable the module in SDK.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam commented:
@MaureenHelm Ping!

raveslave commented:
would be awesome to get this in!

Mani-Sadhasivam commented:
@MaureenHelm ping!

stephanosio commented:
> -:110: WARNING:SPDX_LICENSE_TAG: Missing or malformed SPDX-License-Identifier tag in line 1
> #110: FILE: boards/arm/mimx8mm_evk/doc/index.rst:1:
> +.. _mimx8mm_evk:

Please ignore these warnings (see #24921)

Kludentwo commented on drivers/serial/uart_mcux_iuart.c:
```diff
@@ -0,0 +1,307 @@
+static int mcux_iuart_irq_rx_ready(struct device *dev)
+{
+	const struct mcux_iuart_config *config = dev->config->config_info;
+	u32_t mask = kUART_RxDataReadyFlag;
```
The kUART_RxDataReadyFlag should be kUART_RxDataReadyEnable

Kludentwo commented on drivers/serial/uart_mcux_iuart.c:
```diff
@@ -0,0 +1,307 @@
+	uart_config.baudRate_Bps = config->baud_rate;
+
+	UART_Init(config->base, &uart_config, clock_freq);
+
```
The irq_config_func is never called and should be called in init.
`#ifdef CONFIG_UART_INTERRUPT_DRIVEN                                                                                                                                                                            
         config->irq_config_func(dev);                                                                                                                                                                           #endif`

bwasim commented on soc/arm/nxp_imx/mimx8mm6_m4/soc.c:
```diff
@@ -0,0 +1,144 @@
+
+#include <dt-bindings/rdc/imx_rdc.h>
+
+/* OSC/PLL is already initialized by ROM and Cortex-A7 (u-boot) */
```
A7 or A53 ?

bwasim commented on soc/arm/nxp_imx/mimx8mm6_m4/linker.ld:
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2014 Wind River Systems, Inc.
```
copy right needs fixing i guess..

bwasim commented on drivers/serial/uart_mcux_iuart.c:
```diff
@@ -0,0 +1,314 @@
+
+struct mcux_iuart_config {
+	UART_Type *base;
+	char *clock_name;
```
This can be made `const` assuming that this will never change during the program life cycle.. 

Making things `const` makes them go to the ROM which helps saving RAM (in MCUs, we generally have more ROM than RAM)..

bwasim commented on drivers/serial/uart_mcux_iuart.c:
```diff
@@ -0,0 +1,314 @@
+static void mcux_iuart_irq_tx_enable(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	u32_t mask = kUART_TxEmptyEnable;
```
We could directly pass `kUART_TxEmptyEnable` to the interrupts enable function instead of creating a local variable.. 

The compiler will do this for us anyways, but still better to remove `mask`..

bwasim commented on drivers/serial/uart_mcux_iuart.c:
```diff
@@ -0,0 +1,314 @@
+static void mcux_iuart_irq_tx_disable(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	u32_t mask = kUART_TxEmptyEnable;
```
Similarly, `mask` can be removed from here..

bwasim replied:
And in other similar functions also..

bwasim commented on dts/arm/nxp/nxp_imx8mm_m4.dtsi:
```diff
@@ -0,0 +1,87 @@
+
+		ccm: ccm@30380000 {
+			compatible = "nxp,imx-ccm";
+			reg = <0x30380000 0x10000>;
```
Should use `DT_SIZE_K` here for consistency..

bwasim commented on soc/arm/nxp_imx/mimx8mm6_m4/soc.c:
```diff
@@ -0,0 +1,144 @@
+}
+
+/* AUDIO PLL1 configuration */
+static const ccm_analog_frac_pll_config_t g_audioPll1Config = {
```
Are all of these PLLs configurations applicable to all boards that use this SoC ? Seems that the divisors might change if the same SoC is used with a different board, which probably makes this board-specific stuff instead of SoC-specific.. Can you confirm ?

MaureenHelm commented on drivers/serial/Kconfig.mcux_iuart:
```diff
@@ -0,0 +1,21 @@
+
+if UART_MCUX_IUART
+
+config UART_MCUX_IUART_4
+	bool "UART 4"
+	help
+	  Enable UART 4.
```
We've recently removed driver instance Kconfigs, please see 5a149dfca697e2716d0e96770be1358c9ec0e618

MaureenHelm commented on west.yml:
```diff
@@ -95,7 +95,7 @@ manifest:
       path: tools/net-tools
     - name: hal_nxp
-      revision: 80a337dc4cc980f80e595971147eb41458b29ae8
+      revision: pull/16/head
```
Still need to resolve. Please see https://github.com/zephyrproject-rtos/hal_nxp/pull/16#issuecomment-618023931

Mani-Sadhasivam replied:
Done!

MaureenHelm replied:
Thanks! The hal PR has been merged, so you can now update this to reference the new SHA.

Mani-Sadhasivam commented:
Ping @bwasim @carlescufi @galak

carlescufi commented:
@MaureenHelm could you please re-review and merge?

Mani-Sadhasivam commented:
@MaureenHelm Ping!

Kludentwo commented on drivers/serial/uart_mcux_iuart.c:
```diff
@@ -0,0 +1,332 @@
+static int mcux_iuart_irq_rx_ready(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	uint32_t mask = kUART_RxDataReadyFlag;
```
The kUART_RxDataReadyFlag should be kUART_RxDataReadyEnable.
I have tested this with the uart shell.

MaureenHelm reviewed (dismissed):
Almost there! Just need to rebase the hal PR and a UART driver fix.

Thanks for your patience, @Mani-Sadhasivam . I'm sorry this has taken so long.

Mani-Sadhasivam commented:
> Almost there! Just need to rebase the hal PR and a UART driver fix.

@MaureenHelm Done!

Mani-Sadhasivam commented:
@MaureenHelm Updated the SHA in `west.yaml`

dleach02 reviewed (approved):
lgtm
